src/adc_mux.cpp

```diff
@@ -0,0 +1,71 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <libhal-soft/adc_mux.hpp>
+
+using namespace hal::literals;
+using namespace std::chrono_literals;
+
+namespace hal {
+
+// free functions
+status set_mux_channel(uint32_t p_position,
+                       std::span<output_pin*> p_signal_pins)
+{
+
+  for (int i = 0; i < static_cast<int>(p_signal_pins.size()); i++) {
+
+    bool value = bool(p_position & (1 << i));
+    HAL_CHECK(p_signal_pins[i]->level(value));
+  }
+  return success();
+}
+
+// implementations for adc_mux_pin inner class
+using adc_mux_pin = adc_multiplexer::adc_mux_pin;
+
+adc_mux_pin::adc_mux_pin(uint8_t p_mux_port, adc_multiplexer* p_mux)
+  : m_mux_port{ p_mux_port }
+  , m_mux{ p_mux } {};
+
+result<adc_mux_pin::read_t> adc_mux_pin::driver_read()
+{
+  return m_mux->read_channel(m_mux_port);
+}
```

Comment:
```suggestion
}

```
NIT: add a space between the comment but remove the line below where it starts.